### ISSUE TITLE
Expose scale bar parameters in GUI

### ISSUE THREAD

**Issue by jo-mueller** [napari/napari#7168]
## 🚀 Feature: GUI for scale bar properties 📏🔢 
Following a discussion in #7164, @psobolewskiPhD, @brisvag and I discussed whether it would be helpful for the general audience to expose the scale bar controls in a handy little GUI (as opposed to having to set everything from the CLI. The interface could be a simple magicgui-created interface like this:

![image](https://github.com/user-attachments/assets/8835892a-861a-4c77-8664-e67977c60e88)

## Motivation

Setting scale bar properties should be easy to do for non cli-affine users. Currently, some properties of the scale bar (location, position, etc) can only be changed from code.

## Pitch

The current scale bar submenu (`View > Scale Bar > ...`) should feature one more entry called `Settings`, which should open a GUI in the viewer from which users can set all relevant properties (`unit`, `font_size`, `length` (after #7167 ), `color`, etc) interactively.

The GUI could be created using simple magicgui syntax, whereas I am not sure whether making `length` an optional parameter that can be enabled/disabled from the GUI and adding a qt colorpicker widget will be very easy in magicgui 🤔 

```python
@magic_factory(
    color={'label': 'Color'},
    fixed_width={'label': 'Fixed Width [unit]'},
    unit={'label': 'Unit'},
    font_size={'label': 'Font Size'},
    position={'choices': ['bottom right', 'bottom left', 'top right', 'top left']},
    auto_call=True
)
def create_scalebar(color: str = 'white',
                    fixed_width: Optional[int] = None,
                    unit: Optional[str] = '',
                    font_size: Optional[int] = 12,
                    position: str = 'bottom right',
                    visible: bool = True):
    viewer.scale_bar.visible = visible
    viewer.scale_bar.color = color
    viewer.scale_bar.fixed_width = fixed_width
    viewer.scale_bar.unit = unit
    viewer.scale_bar.font_size = font_size
    viewer.scale_bar.position = position
```

@jni I have seen that you tried something like this (i.e., adding a colorpicker to a magicgui widget [in the past](https://github.com/pyapp-kit/magicgui/issues/490) - is that something that's easily done? Or would you think that it may be asier to simply create the widget in Qt?

## Alternatives

It would probably be possible to add all of the properties as separate entries in the dropdown behind `View > Scale Bar > ...`, but I would find that a bit clunky.

## Additional context

#7167 will bring a feature to control the length of the scale bar to a fixed physical size, so a possible PR adressing this issue should go after that.

**Comment by brisvag** [comment]
> The GUI could be created using simple magicgui syntax, whereas I am not sure whether making `length` an optional parameter that can be enabled/disabled from the GUI and adding a qt colorpicker widget will be very easy in magicgui 🤔

I thought there was a colorpicker in superqt... but I don't see it now. @Czaki do you know if this code is somehwere and can maybe even be added to magicgui? @tlambert03 

You can mark magicgui parameters as `nullable`, which means they may be set to `None`. However, a quick test with a `SpinBox` seems to be broken, which looks like a bug:

```py
from typing import Optional
from magicgui import magicgui

@magicgui(path={'nullable': True})
def f(path: Optional[int] = None):
    print(path, type(path))
f.path.value = None
f.show(run=True)
```

```py
TypeError: float() argument must be a string or a real number, not 'NoneType'
```

cc @tlambert03

**Comment by Czaki** [comment]
`QColorDialog` is part of Qt itself. `from qtpy.QtWidgets import QColorDialog` 

To add it to magicgui it will require to wrap it in proper widget inheritance. I think that @tlambert03 could point us how to do this.

**Comment by jni** [comment]
Talley already added documentation about this at my request, here:

https://pyapp-kit.github.io/magicgui/CONTRIBUTING/#adding-a-widget

**Comment by jni** [comment]
For the type to use, Talley's cmap.ColorLike is probably a good candidate:

https://cmap-docs.readthedocs.io/en/latest/colors/

**Comment by tlambert03** [comment]
There is also a superqt widget 

https://pyapp-kit.github.io/superqt/widgets/qcolorcombobox/

**Comment by Czaki** [comment]
In general, we contribute time to time to `superqt` if we think that widget is enough general that other people may want to use it.

**Comment by jo-mueller** [comment]
I could try myself at an implementation but the question to resolve first would be: Should a scale bar parameter GUI be created 

- automatically through magicgui (requires adding the colorpicker as an extension there first)
- "manually" as a qt/superqt widget?  

I could see the later be a bit easier to realize.

**Comment by jni** [comment]
Generally, I prefer magicgui, because in the long term, it would be good for napari to be portable across frameworks, and not be tied to Qt (specifically, because Qt has no web frontend). So I think we should aim for magicgui going forward (which could then be updated to provide alternative UI frontends).

**Comment by brisvag** [comment]
Agree, and it seems like this specific case should be relatively easy to do with magicgui, given also the widget above :)

I *think* that what you need to do is to register the Color type (https://cmap-docs.readthedocs.io/en/latest/colors/) with magicgui for the superqt colorpicker widget ([pyapp-kit.github.io/superqt/widgets/qcolorcombobox](https://pyapp-kit.github.io/superqt/widgets/qcolorcombobox/) and it should work when you do:

```py
@magicgui
def stuff(color: cmap.Color):
	...
```

**Comment by jni** [comment]
> magicgui for the superqt colorpicker widget ([pyapp-kit.github.io/superqt/widgets/qcolorcombobox](https://pyapp-kit.github.io/superqt/widgets/qcolorcombobox/)

No, I think it should map to an abstract ColorPicker that is implemented as a QColorPicker. As far as I can tell, the QColorComboBox only lets you choose from a pre-determined list of colors, which might be ok for an initial implementation but is not the right approach in general.

**Comment by brisvag** [comment]
> No, I think it should map to an abstract ColorPicker that is implemented as a QColorPicker. As far as I can tell, the QColorComboBox only lets you choose from a pre-determined list of colors, which might be ok for an initial implementation but is not the right approach in general.

Nope, it's much better than that, you can add colors to the dropdown and it spawns a colorpicker to do so!

```py
from magicgui import magicgui
from superqt import QColorComboBox
w = QColorComboBox(allow_user_colors=True)
w.show()
```

![image](https://github.com/user-attachments/assets/895688b1-196a-4a57-b70c-0dd5a03451da)

**Comment by jni** [comment]
Very cool and interesting! Should have read the full page. 😅

I don't know whether you can associate to widgets with non-default parameters though... 🤔

**Comment by psobolewskiPhD** [comment]
Just cross linking here:
https://github.com/napari/napari/issues/7555
Using cmap could be worth it in image colormaps, so that users can access the broader catalog.

**Comment by TimMonko** [comment]
We have currently been adding popup widgets to viewer buttons, see the grid and 2D/3D button. This may be a good place to add such a widget, rather than adding a widget around the canvas. However, this would be best served as a new viewer button (to turn scale_bar on and off, presumably). This is good all around to work on for a sprint, I think!

**Comment by psobolewskiPhD** [comment]
I'm a bit fearful of proliferation of buttons 😬
I feel like buttons for things that are commonly used make sense and then attaching a pop-up to that thing is fine.
But for things that are more rarely used, we have menus and the command palette, so using a widget to enter stuff still feels best to me.

**Comment by jni** [comment]
I agree with you in the general sense @psobolewskiPhD but the scale bar could possibly fit under a right click menu for a canvas/multicanvas button, while moving transpose into the dims order menu. Having said that, at that point it becomes a big design discussion more than a sprint topic. 😅 

However, adding a magicgui-based widget to the builtins would be a nice sprint topic indeed.

**Comment by jo-mueller** [comment]
Thanks for bumping this back up. Last time this was discussed I think the discussion got stuck a bit on the question on where or how to implement the color of the scae bar, with the result that that should best be implemented in magicgui itself. As for the napari side of things, maybe a good place to start would be to expose a simple widget for the other parameters which magicui already supports.

I remember wondering on where exactly that would have to be implemented...is builtins the right place, then?

**Comment by psobolewskiPhD** [comment]
@DragaDoncila if this (hypothetical) widget is added to builtins, does that preclude putting it in View > Scale bar sub-menu? Because it absolutely belongs there IMO...

**Comment by jni** [comment]
Fantastic Q @psobolewskiPhD! You wouldn't be able to do it with NAP-6, but one potential solution is to have the napari code check for built-ins, and if they are present, add it to the menu. (Another solution would be to expand the number of contributable menus, but that might be more contentious.)

**Comment by DragaDoncila** [comment]
Yeah if it's declared via our `builtins` plugin it will only have access to the contributable menu, which doesn't currently include `View`. 

This doesn't really seem to me to a plugin thing, and I don't see why it shouldn't just go in core, but I assume that can of worms has already been thoroughly kicked around.

Outside of that, our options are to expand the list of menus to include `View`, or to special-case our own builtins. Last time we had this discussion (about the measure action I think?), folks didn't want to special-case builtins but it seems valid to me that our core plugin has more access than others.

It's not clear to me at all what the distinction is or should be between our `core` functionality and what goes in `builtins`.

**Comment by psobolewskiPhD** [comment]
My preference is for core, because I think this is a long overdue core thing, but I guess one way to achieve napari lite would be to devolve as much of the GUI functionality to builtins and then make builtins like console, an optional thing.

**Comment by jo-mueller** [comment]
Core then. I think the only thing missing for me to send a PR would be a small hint as to where *exactly* in the code this would have to be implemented 🙊

**Comment by TimMonko** [comment]
> feel like buttons for things that are commonly used make sense and then attaching a pop-up to that thing is fine.
> But for things that are more rarely used, we have menus and the command palette, so using a widget to enter stuff still feels best to me.

This discussion went quickly, but I wanted to circle back to this. How do we decide on a frequently used button? Some users will never touch the console button, and that serves only a single function. The transpose button and reset view buttons also have the same, very limited use case. Why should we limit small buttons in our GUI going forward? Especially if these buttons can open up a whole world of clean GUI design, with useful, transient user oriented tools. 

Each overlay can (and I think should) be served by a button with a popup (axes, text, bounding box, etc). Otherwise we spend so much time making cool API features that most users will never find, and/or are too clunky to use. Like, even toggling on the axes labels could be added to the dims popup where it can already be re-named. 

But, this discussion isn't really the point of this issue. However, I DO think it is something that we, as a community, need to actually decide rather than say 'let it be a widget in the sidebar'. This imo is very bad design as many _gui_ users already have their sidebars difficult to manage from using so many plugins / other widgets. Adding more widget burden over there will be a horrible experience. A button with a popup would be much less intrusive, and live only temporarily in the napari space (a good thing!) 

Anyways, happy to open a bigger discussion if this isn't something that will already go ignored.
